### PR TITLE
feat(pipelines): add control over underlying CodePipeline

### LINKED ISSUE #9021: [Pipelines] Running additional actions in the default "Build" stage
Possibility to run additional actions during the default `Build` phase. Apart from the CDK build, other builds such as website builds might be required. A website might be built and deployed along with the CDK app within the same pipeline.

### Use Case

Running additional build actions in the default `Build` phase such as website builds.

### Proposed Solution

~~Making the build stage a public property of `pipelines.CdkPipeline` so that custom actions can be added.~~

Let the user specify an array of `buildActions` instead of a single `synthAction`, e.g.

```ts
const synthAction = pipelines.SimpleSynthAction.standardYarnSynth({
  sourceArtifact,
  cloudAssemblyArtifact,
});

const pipeline = new pipelines.CdkPipeline(this, 'Pipeline', {
  pipelineName: 'MyAppPipeline',
  cloudAssemblyArtifact,
  sourceAction,
  buildActions: [synthAction, /* more actions */],
});
```

I think this covers all possible needs. Also, it is consistent with `codepipeline.StageProps.actions`:

https://github.com/aws/aws-cdk/blob/403d65735eda8adc82ef6361b1aabae4bb640703/packages/%40aws-cdk/aws-codepipeline/lib/pipeline.ts#L47-L51

### Other

* [x] :wave: I may be able to implement this feature request
* [x] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by rix0rrr**
For people with specific requirements:

* Allow supplying an existing (preconfigured) CodePipeline object,
  via the `codePipeline` argument. This pipeline may already have
  Source and Build stages, in which case `sourceAction` and
  `synthAction` are no longer required.
* Allow access to the underlying CodePipeline object via the
  `.codePipeline` property, and allow modifying it via
  `pipeline.stage("Source").addAction(...)`.

Fixes #9021.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/pipelines/lib/pipeline.ts [hunk 1]
```diff
@@ -12,22 +12,37 @@ import { AddStageOptions, AssetPublishingCommand, CdkStage, StackOutput } from '
...
-  readonly synthAction: codepipeline.IAction;
+  readonly synthAction?: codepipeline.IAction;
 
   /**
    * The artifact you have defined to be the artifact to hold the cloudAssemblyArtifact for the synth action
    */
   readonly cloudAssemblyArtifact: codepipeline.Artifact;
 
+  /**
+   * Existing CodePipeline to modify
```
Comment (nija-at) [inline packages/@aws-cdk/pipelines/lib/pipeline.ts:33]: to modify?

rephrase to something clearer maybe

File: packages/@aws-cdk/pipelines/lib/pipeline.ts [hunk 2]
```diff
@@ -72,28 +87,58 @@ export class CdkPipeline extends Construct {
...
+        restartExecutionOnUpdate: true,
+      });
+    }
+
+    if (props.sourceAction && !props.synthAction) {
+      // Because of ordering limitations, you can: bring your own Source, bring your own
+      // Both, or bring your own Nothing. You cannot bring your own Build (which because of the
+      // current CodePipeline API must go BEFORE what we're adding) and then having us add a
+      // Source after it. That doesn't make any sense.
+      throw new Error('When passing a \'sourceAction\' you must also pass a \'synthAction\' (or a \'codePipeline\' that already has both)');
```
Comment (nija-at) [inline packages/@aws-cdk/pipelines/lib/pipeline.ts:108]: I personally find the escaped quotes annoying to read. I use quote-in-quote like - 'When passing a "sourceAction" ...'

File: packages/@aws-cdk/pipelines/lib/pipeline.ts [hunk 3]
```diff
@@ -112,10 +157,19 @@ export class CdkPipeline extends Construct {
    * You can use this to add more Stages to the pipeline, or Actions
    * to Stages.
    */
-  public get pipeline(): codepipeline.Pipeline {
+  public get codePipeline(): codepipeline.Pipeline {
```
Comment (nija-at) [inline packages/@aws-cdk/pipelines/lib/pipeline.ts:160]: call out as a breaking change?

**Review by nija-at — APPROVED** [review]

File: packages/@aws-cdk/pipelines/lib/pipeline.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/pipelines/lib/pipeline.ts:108]: I agree, but I definitely WANT single quotes in the string that ultimately gets displayed (those are a lot lighter on the eyes than double quotes, imo), and the rest is forced upon me by our linter.

File: packages/@aws-cdk/pipelines/lib/pipeline.ts (hunk 3 above)
Comment (rix0rrr) [reply packages/@aws-cdk/pipelines/lib/pipeline.ts:160]: This sucks. The previous one was committed by accident. I believe there hasn't been a release in between.

**Pull request merged**